python/ray/tune/schedulers/pbt.py

```diff
@@ -59,46 +60,59 @@ def _explore(
             in the PopulationBasedTraining scheduler.
         resample_probability: Probability of allowing resampling of a
             particular variable.
+        perturbation_factors: Scaling factors to choose between when mutating
+            a continuous hyperparameter.
         custom_explore_fn: Custom explore fn applied after built-in
             config perturbations are.
+
+    Returns:
+        new_config: New hyperparameter configuration (after random mutations).
+        operations: Map of hyperparam -> string describing mutation operation performed
     """
+    operations = {}
     new_config = copy.deepcopy(config)
     for key, distribution in mutations.items():
         if isinstance(distribution, dict):
-            new_config.update(
-                {key: _explore(config[key], mutations[key], resample_probability, None)}
+            nested_new_config, nested_ops = _explore(
+                config[key],
+                mutations[key],
+                resample_probability,
+                perturbation_factors,
+                custom_explore_fn=None,
             )
+            new_config.update({key: nested_new_config})
+            operations.update({key: nested_ops})
         elif isinstance(distribution, list):
             if (
                 random.random() < resample_probability
                 or config[key] not in distribution
             ):
                 new_config[key] = random.choice(distribution)
-            elif random.random() > 0.5:
-                new_config[key] = distribution[
-                    max(0, distribution.index(config[key]) - 1)
-                ]
+                operations[key] = "resample"
             else:
-                new_config[key] = distribution[
-                    min(len(distribution) - 1, distribution.index(config[key]) + 1)
-                ]
+                shift = random.choice([-1, 1])
+                new_idx = distribution.index(config[key]) + shift
+                new_idx = min(max(new_idx, 0), len(distribution) - 1)
+                new_config[key] = distribution[new_idx]
+                operations[key] = f"shift {'left' if shift == -1 else 'right'}"
```

Comment:
Should this be "shift left (noop)" or similar if old_idx == new_idx? E.g. when we select shift = -1 when we're already at the first item

Reply:
Right, this should include some indicator that we've hit the end of the list.

A few thoughts on this actually:
- We might want to just wrap around from the end if trying to shift left at the first item. The fact that we don't wrap implies some kind of ordering to the items, but there is no guarantee there. Example:
    - `"a": [100, 1, 50, 75]`
    - `75 <- (shift left & wrap around) -- 100 -- (shift right) -> 1`
    - Seems arbitrary why 100 can only be perturbed to the right.
- Should a `noop` be an option here rather than always shifting left and right? [SHERPA's implementation](https://github.com/sherpa-ai/sherpa/blob/ff6466e99717983f9f394ba72b63f17343e32bdc/sherpa/algorithms/core.py#L717) includes a "no shift" option.

Reply:
Hm what does the original paper say?

Everything else seems to have an option not to shift, so I actually believe just going with [-1, 0, 1] should be fine.

For wrapping around, basically a matter of preference. Best would be to distinguish between ordinal and categorical, but since we don't have this, I'd say let's assume ordinal and leave the logic as is

Reply:
1. The original paper doesn't really mention specifics on perturb, so I think adding the no-shift makes sense to align with other implementations.

2. I see, can leave this for a PR in the future.